src/actions/progress.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use actions::notifications::{PublishDiagnostics, Progress};
+use lsp_data::{ProgressParams, PublishDiagnosticsParams};
+use server::{Output, Notification};
+
+/// Trait for communication of build progress back to the client.
+pub trait ProgressNotifier: Send {
+    fn notify_begin_progress(&self);
+    fn notify_progress(&self, update: ProgressUpdate);
+    fn notify_end_progress(&self);
+}
+
+/// Kinds of progress updates
+pub enum ProgressUpdate {
+    Message(String),
+    Percentage(f64),
+}
+
+/// Trait for communication of diagnostics (i.e. build results) back to the rest of
+/// the RLS (and on to the client).
+// This trait only really exists to work around the object safety rules (Output
+// is not object-safe).
+pub trait DiagnosticsNotifier: Send {
```

Comment:
One idea to deduplicate the traits and impl logic might be to introduce a RAII guard that both ensures that we have an active `windows/progress` session that we finish it with `"done": true`, which also exposes a way to report specific progress for it (be it publishing diagnostics or sending a progress message), but that'd be mostly exploratory - I'm fine with what we currently have here.